src/RandMsg.chpl

```diff
@@ -547,6 +548,101 @@ module RandMsg
         return new MsgTuple(repMsg, MsgType.NORMAL);
     }
 
+    proc poissonGeneratorMsg(cmd: string, msgArgs: borrowed MessageArgs, st: borrowed SymTab): MsgTuple throws {
+        const pn = Reflection.getRoutineName(),
+              name = msgArgs.getValueOf("name"),                                // generator name
+              isSingleLam = msgArgs.get("is_single_lambda").getBoolValue(),     // boolean indicated if lambda is a single value or array
+              lamStr = msgArgs.getValueOf("lam"),                               // lambda for poisson distribution
+              size = msgArgs.get("size").getIntValue(),                         // number of values to be generated
+              state = msgArgs.get("state").getIntValue(),                       // rng state
+              rname = st.nextName();
+
+
+        randLogger.debug(getModuleName(),getRoutineName(),getLineNumber(),
+                                "name: %? size %i isSingleLam %? lamStr %? state %i".doFormat(name, size, isSingleLam, lamStr, state));
+
+        st.checkTable(name);
+
+        var generatorEntry: borrowed GeneratorSymEntry(real) = toGeneratorSymEntry(st.lookup(name), real);
+        ref rng = generatorEntry.generator;
+        if state != 1 {
+            // you have to skip to one before where you want to be
+            rng.skipTo(state-1);
+        }
+
+        // uses the algorithm from knuth found here:
+        // https://en.wikipedia.org/wiki/Poisson_distribution#Random_variate_generation
+        // generates values drawn from poisson distribution using a stream of uniformly distributed random numbers
+
+        const nTasksPerLoc = here.maxTaskPar; // tasks per locale based on locale0
+        const Tasks = {0..#nTasksPerLoc}; // these need to be const for comms/performance reasons
+
+        const generatorSeed = (rng.next() * 2**62):int;
+        var poissonArr = makeDistArray(size, int);
+
+        // I hate the code duplication here but it's not immediately obvious to me how to avoid it
```

Comment:
Not sure if it's worth all the overhead, but you could use a record type that can store either a single lambda value, or an array of lambdas s.t. `const lam = lamArr[locSubDom.low + i];` always works:
```chapel
record lambda {
  var isArray: bool;
  var arr;
  var val: real;

  proc init(arr: [] real) {
    this.isArray = true;
    this.arr = arr;
  }

  proc init(val) {
    this.isArray = false;
    this.arr = makeDistArray(0..0);
    this.val = val:real;
  }

  proc this(idx) {
    return if isArray then this.arr[idx] else this.val;
  }
}
```

And then initialize it differently depending on the value of `isSingleLam`:
```chapel
var lamArr = if isSingleLam then new lambda(lamStr)
  else new lambda(toSymEntry(getGenericTypedArrayEntry(lamStr, st),real).a);
```

Reply:
oh yah I like this a lot! thanks for this! I bet theres quite a few places in the codebase that benefit from something like this